couchbase/src/bucket.rs

```diff
@@ -33,28 +33,28 @@ impl Bucket {
 
     /// Internal proxy method that gets called from the cluster so we can send it into the
     /// instance.
-    pub(crate) fn query<S>(
+    pub(crate) async fn query<S>(
         &self,
         statement: S,
         options: Option<QueryOptions>,
-    ) -> impl Future<Item = QueryResult, Error = CouchbaseError>
+    ) -> Result<QueryResult, CouchbaseError>
     where
         S: Into<String>,
     {
-        self.instance.query(statement.into(), options)
+        self.instance.query(statement.into(), options).await
     }
 
     /// Internal proxy method that gets called from the cluster so we can send it into the
     /// instance.
-    pub(crate) fn analytics_query<S>(
+    pub(crate) async fn analytics_query<S>(
```

Comment:
There are many methods like this so I'll just take this one to discuss the pattern:

I'm unsure if these "forward" methods  gain anything from being converted to async await, since we are not actually awaiting on something? I'm somewhat cautious that the async/await sugar for these method introduces some (unnecessary?) overhead that is not needed since it is just forwarding the underlying future result?

On the other hand I have no data to back that feeling up and I am also not sure what the best practice is going to be on that. What's your take on this?

Reply:
Other than that, looks pretty great to me :)

Reply:
I struggled with this! I'm not an expert on async/await - but here is why I landed where I did. Essentially, when you declare a function as `async`, what Rust does is generate the future state machine for you, based on all the calls to `await` within the function. 

Compared to futures 0.1, there is a little more overhead (you can see this in some of the outstanding issues around async/await - in particular, things can get a little out of control if you pass futures as arguments to functions, or worse, if you do it recursively). I'm pretty sure that it won't be a big differential, either in performance or resource utilization. 

As a bit of practicality, when I "passed through" the future generated from the real function, rather than the forward one, I wound up with an internal API that had a lot of calls to `await.await`- which felt like a code smell. In an async/await world, I think the right thing to do is let the compiler do the work.

Reply:
In practice, these functions are async - they return the future from upstream when it is available. Nothing actually happens, of course, until they are passed to an executor.

Reply:
I checked in with the tokio discord - there shouldn't be an advantage to returning the raw future, unless the intent is to build the state machine by hand (which we certainly don't want).

Reply:
@adamhjk thanks much for checking, looks good to me then!